File: backend/app/api/v1/tags.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy import func
from typing import List, Optional
from uuid import UUID

from app.core.database import get_db
from app.models.tag import Tag, post_tags
from app.models.post import Post, PostStatus
from app.schemas.tag import TagResponse, TagWithCount

router = APIRouter(prefix="/tags", tags=["tags"])

@router.get("/hot", response_model=List[TagWithCount])
async def get_hot_tags(limit: int = Query(10, ge=1, le=50), db: AsyncSession = Depends(get_db)):
    # Join with post_tags and posts (where is_published == True and status == approved)
    stmt = (
        select(Tag, func.count(post_tags.c.post_id).label("post_count"))
        .outerjoin(post_tags, post_tags.c.tag_id == Tag.id)
        .outerjoin(Post, (Post.id == post_tags.c.post_id) & (Post.is_published == True) & (Post.status == PostStatus.APPROVED))
        .group_by(Tag.id)
        .order_by(func.count(post_tags.c.post_id).desc(), Tag.name.asc())
        .limit(limit)
    )
    result = await db.execute(stmt)
    tags_with_counts = []
    for row in result.all():
        tag, count = row[0], row[1]
        tags_with_counts.append(TagWithCount(
            id=tag.id,
            name=tag.name,
            slug=tag.slug,
            created_at=tag.created_at,
            post_count=count
        ))
    return tags_with_counts

@router.get("/search", response_model=List[TagResponse])
async def search_tags(q: str = Query("", min_length=0), limit: int = Query(20, ge=1, le=50), db: AsyncSession = Depends(get_db)):
    if not q.strip():
        stmt = select(Tag).order_by(Tag.name.asc()).limit(limit)
    else:
        pattern = f"%{q.strip()}%"
        stmt = select(Tag).filter((Tag.name.ilike(pattern)) | (Tag.slug.ilike(pattern))).order_by(Tag.name.asc()).limit(limit)
    result = await db.execute(stmt)
    return result.scalars().all()

@router.get("", response_model=List[TagResponse])
async def list_tags(skip: int = 0, limit: int = 100, db: AsyncSession = Depends(get_db)):
    stmt = select(Tag).order_by(Tag.name.asc()).offset(skip).limit(limit)
    result = await db.execute(stmt)
    return result.scalars().all()

@router.get("/{id_or_slug}", response_model=TagWithCount)
async def get_tag(id_or_slug: str, db: AsyncSession = Depends(get_db)):
    try:
        tag_id = UUID(id_or_slug)
        stmt = (
            select(Tag, func.count(post_tags.c.post_id).label("post_count"))
            .outerjoin(post_tags, post_tags.c.tag_id == Tag.id)
            .outerjoin(Post, (Post.id == post_tags.c.post_id) & (Post.is_published == True) & (Post.status == PostStatus.APPROVED))
            .filter(Tag.id == tag_id)
            .group_by(Tag.id)
        )
    except ValueError:
        stmt = (
            select(Tag, func.count(post_tags.c.post_id).label("post_count"))
            .outerjoin(post_tags, post_tags.c.tag_id == Tag.id)
            .outerjoin(Post, (Post.id == post_tags.c.post_id) & (Post.is_published == True) & (Post.status == PostStatus.APPROVED))
            .filter(Tag.slug == id_or_slug)
            .group_by(Tag.id)
        )

    result = await db.execute(stmt)
    row = result.first()
    if not row:
        raise HTTPException(status_code=404, detail="Tag not found")
    tag, count = row[0], row[1]
    return TagWithCount(
        id=tag.id,
        name=tag.name,
        slug=tag.slug,
        created_at=tag.created_at,
        post_count=count
    )

File: backend/src/routes/tags.ts
import { Hono } from 'hono';
import { and, asc, count, desc, eq, ilike, or } from 'drizzle-orm';
import { db } from '../db/index.js';
import { postTags, posts, tags } from '../db/schema.js';
import { notFound } from '../core/errors.js';
import { asUuid } from '../core/security.js';
import { toTagResponse, toTagWithCount } from '../schemas/responses.js';

export const tagRoutes = new Hono();

/**
 * The post_count carried by /tags/hot and /tags/{id_or_slug} counts rows in
 * post_tags, not approved posts: tags.py left-joins posts with a status
 * filter but then counts post_tags.post_id, which the outer join cannot
 * reduce. Reproduced as-is so the numbers in the sidebar do not move.
 */
const linkCountJoins = {
  postTagsOn: eq(postTags.tag_id, tags.id),
  postsOn: and(
    eq(posts.id, postTags.post_id),
    eq(posts.is_published, true),
    eq(posts.status, 'approved'),
  ),
};

// /hot and /search must be registered before /:id_or_slug, or they would be
// swallowed by the parameter route.
tagRoutes.get('/hot', async (c) => {
  const limitRaw = Number(c.req.query('limit') ?? 10);
  const limit = Math.min(50, Math.max(1, Number.isFinite(limitRaw) ? limitRaw : 10));

  const rows = await db
    .select({ tag: tags, post_count: count(postTags.post_id) })
    .from(tags)
    .leftJoin(postTags, linkCountJoins.postTagsOn)
    .leftJoin(posts, linkCountJoins.postsOn)
    .groupBy(tags.id)
    .orderBy(desc(count(postTags.post_id)), asc(tags.name))
    .limit(limit);

  return c.json(rows.map((r) => toTagWithCount(r.tag, Number(r.post_count))));
});

tagRoutes.get('/search', async (c) => {
  const q = (c.req.query('q') ?? '').trim();
  const limitRaw = Number(c.req.query('limit') ?? 20);
  const limit = Math.min(50, Math.max(1, Number.isFinite(limitRaw) ? limitRaw : 20));

  const rows = q
    ? await db
        .select()
        .from(tags)
        .where(or(ilike(tags.name, `%${q}%`), ilike(tags.slug, `%${q}%`)))
        .orderBy(asc(tags.name))
        .limit(limit)
    : await db.select().from(tags).orderBy(asc(tags.name)).limit(limit);

  return c.json(rows.map(toTagResponse));
});

tagRoutes.get('/', async (c) => {
  const skip = Math.max(0, Number(c.req.query('skip') ?? 0) || 0);
  const limitRaw = Number(c.req.query('limit') ?? 100);
  const limit = Math.min(500, Math.max(1, Number.isFinite(limitRaw) ? limitRaw : 100));

  const rows = await db.select().from(tags).orderBy(asc(tags.name)).offset(skip).limit(limit);
  return c.json(rows.map(toTagResponse));
});

tagRoutes.get('/:id_or_slug', async (c) => {
  const key = c.req.param('id_or_slug');
  const id = asUuid(key);

  const rows = await db
    .select({ tag: tags, post_count: count(postTags.post_id) })
    .from(tags)
    .leftJoin(postTags, linkCountJoins.postTagsOn)
    .leftJoin(posts, linkCountJoins.postsOn)
    .where(id ? eq(tags.id, id) : eq(tags.slug, key))
    .groupBy(tags.id)
    .limit(1);

  const row = rows[0];
  if (!row) throw notFound('Tag not found');
  return c.json(toTagWithCount(row.tag, Number(row.post_count)));
});
